INET 4 compatible version of simulte

wischhof opened the pull request:
In order to be able to use INET 4, significant changes had to be made to the simulte sources, including but not limited to:

- adaptation of UE, eNB, SGW, PGW to new node structure
- adapt all parts on network layer and above to use inet::Packet instead of cPacket
- using MessageDispatchers between the individual layers (->ProtocolTags, InterfaceReqs)
- update all ini files to use the new parameter names
- module initialization
- adaptations to new socket and application interfaces
- use new INET 4 class names 

In total, almost all source code files had to be touched - therfore, this is a quite large pull request. All simulations within the simulation folder (including the cars scenario which makes use of veins) are runnable. However, an in depth-comparison of the simulation results with those obtained with the INET 3 based version has not yet been done.

wischhof commented on .github/workflows/build-and-test.yml:
```diff
@@ -12,5 +12,5 @@ jobs:
       run: |
         docker run --rm -v ${{ github.workspace }}:/root/models -u "$(id -u):$(id -g)" \
-        omnetpp/inet:o5.5.1-3.6.6 \
+        omnetpp/inet:o5.5.1-4.1.1 \
```
We would need a Docker container with Omnet 5.5.1 and INET 4.1.1.  Seems that guessing the name does not work - what would be the correct name?

torokati44 replied:
The omnetpp/inet:o5.5.1-4.1.1 image has now been built.
Any references to the /root/inet/ directory in it should be updated to /root/inet4/ for the new image.

torokati44 replied:
Or, if that would be too much trouble (an inconvenient difference from the development setups); perhaps a soft or hard link named `inet` could be created to `inet4` in the container as an early step of the test.

rhornig replied:
Inside the top level makefile, you should change `-KINET_PROJ=../../inet` to `-KINET_PROJ=../../inet4`

wischhof replied:
Thanks for setting up the container - changed the path to the INET project to ../inet4. Now the project is found.

wischhof replied:
Fingerprint tests were failing since we used a more current version of INET4 in our CI. However, we have recalculated the fingerprints based on INET4 Tag v4.1.1 now and tests pass.
@rhornig Regarding the mistakenly committed files you were right, of course these should never have been committed. Have been removed, thanks for noticing.

rhornig commented:
I'm seeing several suspicious big files in `tests/fingerprint` like: argpars, copy, csv, glob, os, re, time, unittest. Those seem to be photoshop images ??? i.e. they are mistakenly committed.

rhornig commented:
Are the fingerprint tests supposed to PASS? A lot of them are failing currently.

levy commented on src/apps/alert/AlertReceiver.cc:
```diff
@@ -45,17 +46,23 @@ void AlertReceiver::handleMessage(cMessage *msg)
-        throw cRuntimeError("AlertReceiver::handleMessage - FATAL! Error when casting to AlertPacket");
+    Packet* pPacket = check_and_cast<Packet*>(msg);
+
+    if (pPacket == 0)
```
check_and_cast cannot return a nullptr, so there's no point in checking the result

wischhof replied:
Has been removed

levy commented on src/apps/alert/AlertReceiver.cc:
```diff
@@ -45,17 +46,23 @@ void AlertReceiver::handleMessage(cMessage *msg)
 
     // emit statistics
-    simtime_t delay = simTime() - pkt->getTimestamp();
+    simtime_t delay = simTime() - alert->getTimestamp();
```
Using getTimestamp for statistics is generally not a god idea, because Packet instances can be copied at any given time in the underlying network. I think you should attach a CreationTimeTag to the packet data instead.

wischhof replied:
Good idea, we will add a CreationTimeTag (not only in the Alert-App but also all other apps).

Howevever, regarding the statistics calculation (i.e. the delay of a received packet within the receiver) my understanding of the API is the following (correct me if I am wrong): If the Packet instance would be copied in the underlying network, the timeStamp field inserted by the application would also be copied, and a _new_ CreationTimeTag might be added. However, since on the application layer we want to calculate the end-to-end delay (time difference between creation of the original packet and the reception at the application layer at receiver side), using the timeStamp inserted by the application would still be correct, wouldn't?

Looking a the other inet protocols, it seems that for calculating the delay based on the CreationTimeTag the protocols (except for the SCTP implementation) do not implement any statistics but rely on ageneric mechanism (DataAgeFilter)?

However, in order not to break existing scripts relying on these statistics reported by simulte,I would suggest the following:

1. We add CreationTimeTags in all simulte applications.
1. BUT we still calculate the application layer delay based on the application layer time stamp field (users could alternatively use the generic filters if they want the delay calculation to be based on the CreationTimeTag.

Would that be ok in your opinion?

levy replied:
The CreationTimeTag attached to the data will be copied with the data transparently in all involved protocols throughout the underlying network. In contrast, the timestamp field of cMessage will not be maintained. In fact, there's no way of maintaining it correctly. Think about fragmentation and aggregation for a TCP connection, the receiver application can receive a packet that contains data which was created at different times in the sending application. There's simply no way of representing that with a simple timestamp field. So no, using the timestamp field of the cMessage is generally not correct. It may produce the same results in some simple networks, but may be completely wrong in others.

DataAgeFilter is the right mechanism for such cases. The MessageTsAgeFilter is not quite right, but what is more important is that the protocols _do not_ copy the timestamp when creating packets. For example, TCP may create a TCP segment packet from multiple packets, so it doesn't even make sense talking about copying the timestamp field.

So for point 2 I would suggest to use the DataAgeFilter instead.

wischhof replied:
I think there is a misunderstanding (due to the maybe not optimal naming of header fields in simulte)  - getTimestamp() is not called on a cMessage, it is called on the packet header of the AlertPacket, which includes an own (application layer) timestamp:
```
class AlertPacket extends inet::FieldsChunk
{
    unsigned int sno;
    simtime_t timestamp;
}
```

levy replied:
Oh, I see. But in this case the original AlertPacket definition _did hide_ the cMessage timestamp field though. The cMessage::getTimestamp() function is not virtual, so calling getTimestamp on a cPacket didn't return the application specific timestamp. One must have been very careful about this.

But you are right, apart from that, my reasoning above doesn't apply. If the application has a specific timestamp field then it's safe to use that.

levy commented on src/apps/cbr/CbrReceiver.cc:
```diff
@@ -54,7 +55,7 @@ void CbrReceiver::handleMessage(cMessage *msg)
     numReceived_++;
     totFrames_ = pPacket->getNframes(); // XXX this value can be written just once
-    int pktSize = (int)pPacket->getByteLength();
+    int pktSize = (int)pPacket->getSize();
```
getSize() is not a Packet function, so I guess this is not a Packet instance. It's suspicious to me that this is not correct. check_and_casting a cMessage to CbrPacket may not be what you want to do here.

wischhof replied:
> check_and_casting a cMessage to CbrPacket may not be what you want to do here
Argh, yes you are right, thanks for spotting that one! 

We'll correct this and afterwards getSize() operates on the size field of the CBR header, which is ok.

(We did not find that bug since currently no simulte simulation within the simulation folder there is no simulation using the CbrReceiver... Therefore, this code is currently untested. I guess it would make sense to either remove this app completely or to add a small test scenario.)

wischhof replied:
We fixed that one and also added two new test scenarios (CBR-UL, CBR-DL) within the "Demo" simulation. Fingerprint tests for those two scenarios were also added.

levy commented on src/apps/cbr/CbrSender.cc:
```diff
@@ -110,15 +118,14 @@ void CbrSender::initTraffic()
+    cbr->setNframes(nframes_);
+    cbr->setIDframe(iDframe_);
+    cbr->setTimestamp(simTime());
+    cbr->setSize(size_);
```
If setSize sets the content of a size field here (i.e. the bits in the packet), then you may consider renaming the field to sizeField (to have a setSizeField). This change makes it easier to differentiate between the length/size field and the actual length/size of the chunk.

wischhof replied:
Good point. I guess the same is true for the timestamp field. Therefore, we renamed the "size" field within all the application layer packets to "payloadSize" and "timestamp" to "payloadTimestamp". This should make the difference more distinguishable.

levy commented on src/apps/d2dMultihop/MultihopD2D.cc:
```diff
@@ -294,28 +319,50 @@ void MultihopD2D::handleTrickleTimer(cMessage* msg)
+    auto src = pPacket->popAtFront<MultihopD2DPacket>();
+
+    // create a relay packet using the data of the source packet
+    auto data = pPacket->popAtFront<ByteCountChunk>();
```
This popAtFront will not work when working with serialized packets, because it implicitly assumes the the representation is actualy a ByteCountChunk, which may or may not what you want to do here. If you want to process all remaining data up to the back pointer of the packet, then you should use peekData(), etc.

wischhof replied:
Was changed to peekData().

levy commented on src/apps/vod/M1Message.msg:
```diff
@@ -11,7 +11,7 @@ cplusplus {{
 }};
 
-struct inet::L3Address;
+class inet::L3Address { @existingClass; }
```
You could simply use an import here, I guess.

wischhof replied:
Yes, you are right. Was changed to an import.

levy commented on src/apps/voip/VoipPacketSerializer.cc:
```diff
@@ -0,0 +1,62 @@
+
+    int64_t remainders = B(voipPacket->getChunkLength() - (stream.getLength() - startPosition)).get();
+    if (remainders < 0)
+        throw cRuntimeError("VoipPacket length = %d smaller than required %d bytes", (int)B(voipPacket->getChunkLength()).get(), (int)B(stream.getLength() - startPosition).get());
```
The next INET version will automatically check if the serialized and non-serialized forms have the exact same length.

levy replied:
You could also use a separate chunk for the data part.

wischhof replied:
For the time being, I left the size check - we can remove it, when the next INET version is available. A separate data chunk would be an option, but right now we would not really benefit.

levy commented on src/common/LteCommon.cc:
```diff
@@ -13,6 +13,14 @@
+using namespace inet;
+
+const inet::Protocol LteProtocol::lteuu("lteuu", "IPv4 (LTE Uu)");
+Register_Protocol_Dissector(&LteProtocol::lteuu, Ipv4ProtocolDissector);
```
Ipv4ProtocolDissector doesn't seem right.

wischhof replied:
Hmm, actually it is right since this is the IPv4 protocol on the Uu interface (we need a different protocol identifier than IPv4 for the traffic filter) - I renamed it to ipv4uu, so that it becomes more obvious that it is IPv4.

levy commented on src/corenetwork/lteip/IP2lte.cc:
```diff
@@ -42,44 +64,47 @@ void IP2lte::initialize(int stage)
+            throw cRuntimeError("unhandled node type: %i", nodeType_);
+        }
+    }
+    else if (stage == inet::INITSTAGE_NETWORK_LAYER - 1) {
```
Hmm, the -1 is a bad idea, you can't rely on that. What is the intention here?

wischhof replied:
This is a left-over from the INET3 version, I guess the intention was to have the statements executed before the network stage. Replaced it with the right init stages.

levy commented on src/corenetwork/lteip/IP2lte.cc:
```diff
@@ -154,18 +192,17 @@ void IP2lte::fromIpUe(IPv4Datagram * datagram)
-    // TODO Add support to IPv6
-    IPv4Address srcAddr  = datagram->getSrcAddress() ,
-                destAddr = datagram->getDestAddress();
+    // TODO Add support to Ipv6
```
Take a look at L3Tools.cc and how the functions are used elsewhere, because it may help support both IPv4 and IPv6.

wischhof replied:
Thanks for the pointer to the L3Tools. However, since there are many places where protocol support is hard-coded to IPv4 and the INET3 version of simulte is also hard-coded to IPv4, I would vote to add IPv6/more generic L3 support at a later point in time within a separate pull request. Added a comment with a reference to the L3Tools.

levy commented on src/corenetwork/lteip/IP2lte.cc:
```diff
@@ -175,27 +212,35 @@ void IP2lte::toStackUe(IPv4Datagram * datagram)
+    switch (transportProtocol) {
+        case IP_PROT_TCP: {
+            auto& tcpHdr = datagram->peekAtFront<tcp::TcpHeader>(b(-1),
+                    Chunk::PF_ALLOW_SERIALIZATION);
```
I think this PF_ALLOW_SERIALIZATION is not needed here. I guess it's just a leftover from the migration process. This flag allows serialization between arbitrary headers. Binary to/from TcpHeader is allowed by default.

wischhof replied:
Yes, correct. Removed the B(-1) and the flag.

levy commented on src/corenetwork/lteip/LteIp.cc:
```diff
@@ -132,195 +135,226 @@ void LteIp::endService(cPacket *msg)
+    header->setTypeOfService(ipControlInfo->getTypeOfService());
+    // TODO: header immutable? may be dropped
+    datagram->popAtFront<Ipv4Header>();
+    datagram->insertAtFront(header);
```
I think you shoud call trimAtFront() before being able to call insertAtFront() after calling popAtFront().

wischhof replied:
Yes, call to popFront() added.

levy commented on src/corenetwork/lteip/LteIp.cc:
```diff
@@ -132,195 +135,226 @@ void LteIp::endService(cPacket *msg)
+    auto protocolId = ipControlInfo->getProtocolId();
+    if (protocolId == IP_PROT_TCP) {
+        inet::tcp::TcpHeader* tcpseg;
+        tcpseg = check_and_cast<inet::tcp::TcpHeader*>(transportPacket);
```
I think transportPacket cannot be cast to a TcpHeader, because it's a packet now a chunk.

wischhof replied:
Yes, this part was not correctly implemented - is not used in the existing simulation scenarios and therefore this was not discovered. Thanks for spotting the problem! We fixed this one and similar ones in the same file.

levy commented on src/corenetwork/lteip/LteIp.cc:
```diff
@@ -132,195 +135,226 @@ void LteIp::endService(cPacket *msg)
-    transportPacket->setControlInfo(controlInfo);
+    //transportPacket->setControlInfo(controlInfo);
+    packet->popAtFront<Ipv4Header>();
+    packet->insertAtFront(Ptr<Ipv4Header>(controlInfo.get()));
```
I guess this is a temporary solution.

wischhof replied:
This was a left-over from the migration process and completely unnecesssary. We removed it completely.

levy commented on src/stack/compManager/compManagerProportional/LteCompManagerProportional.cc:
```diff
@@ -157,18 +160,20 @@ X2CompProportionalReplyIE* LteCompManagerProportional::buildCoordinatorReply(X2N
 }
 
-void LteCompManagerProportional::handleClientRequest(X2CompMsg* compMsg)
+void LteCompManagerProportional::handleClientRequest(inet::IntrusivePtr<X2CompMsg> compMsg)
```
Using Ptr<X2CompMsg> allows using both kinds of shared pointers.

wischhof replied:
Was changed to Ptr.

levy commented on src/stack/phy/layer/LtePhyRelay.cc:
```diff
@@ -27,7 +29,7 @@ void LtePhyRelay::initialize(int stage)
         WATCH(masterId_);
     }
-    else if (stage == 1)
+    else if (stage == inet::INITSTAGE_PHYSICAL_ENVIRONMENT)
```
This stage is for the physical environment such as physical objects, ground model, etc. There's another one for the physical layer called INITSTAGE_PHYSICAL_LAYER.

wischhof replied:
Was changed to INITSTAGE_PHYSICAL_LAYER.

levy commented on src/x2/LteX2Manager.cc:
```diff
@@ -96,66 +104,60 @@ void LteX2Manager::handleMessage(cMessage *msg)
+        X2NodeId targetEnb = *it;
+        auto pktDuplicate = pkt->dup();
+        auto updatedX2Msg = pktDuplicate->removeAtFront<LteX2Message>();
+        updatedX2Msg->markMutableIfExclusivelyOwned();
```
The removed chunk is mutable by default, you don't have to call the markMutable... function.

wischhof replied:
Call to markMutable was removed.

levy commented:
I reviewed the pull request from the INET API usage point of view and added my comments inline. I think this is a big step in the right direction.

wischhof commented:
@levy @rhornig I read though your comments - very helpful! - and see the point in most of them. We will be checking them in the next days and update the pull request, I'll drop a note when we have solved all the issues.

wischhof commented:
We have addressed all of the issues/comments and updated the pull request accordingly. 

Most were minor changes. However, in two cases (namely the CbrSender/CbrReceiver apps and the LteIp module) the problems you spotted were within previously untested code. (Seems that these parts are also not used within any of the simulations in the INET3 version of simulte.) Therefore, besides fixing the problems we added two additional scenarios within the Demo simulation and the respective fingerprints. Thanks for your detailed review and dentifiying those problems!

levy commented:
That's great!

wischhof commented:
@rhornig @levy @giovanninardini @kruviser What is your opinion on how we should proceed? Since there are many significant changes required for INET4 and currently I have no clue, if and how many other people have tested the simulations besides us, one intermediate step could be pull not on the master branch but maybe on a separeate inet4-Branch?  

We could mention that branch within the README of the master and/or the simuLTE website so that we have hopefully more people using and testing it.

kruviser commented:
I have no objection in this respect. We are following the discussion from
the background but we haven't tested your work yet.


Il giorno gio 21 nov 2019 alle ore 10:01 wischhof <notifications@github.com>
ha scritto:

> @rhornig <https://github.com/rhornig> @levy <https://github.com/levy>
> @giovanninardini <https://github.com/giovanninardini> @kruviser
> <https://github.com/kruviser> What is your opinion on how we should
> proceed? Since there are many significant changes required for INET4 and
> currently I have no clue, if and how many other people have tested the
> simulations besides us, one intermediate step could be pull not on the
> master branch but maybe on a separeate inet4-Branch?
>
> We could mention that branch within the README of the master and/or the
> simuLTE website so that we have hopefully more people using and testing it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/inet-framework/simulte/pull/43?email_source=notifications&email_token=AAMPQJRXANFPLCJETNLBE53QUZE45A5CNFSM4JNCH2K2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEZO4VQ#issuecomment-556985942>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMPQJRMTY6STKIA472S7OTQUZE45ANCNFSM4JNCH2KQ>
> .
>


-- 
Antonio Virdis

rhornig commented:
Agreed. I have created an 'inet4-porting' branch. Please send a pull request against that and I will publish it and amend the main README file on the master branch.

wischhof commented:
> 
> 
> Agreed. I have created an 'inet4-porting' branch. Please send a pull request against that and I will publish it and amend the main README file on the master branch.

I have sent a new pull request against that branch. I guess we could now close/cancel this one, right?
